## Segmenting and Clustering Neighborhoods in Toronto

In [2]:
import numpy as np
import pandas as pd
from bs4 import BeautifulSoup
from urllib.request import urlopen

## Q.1
#### Scraping the Wikipedia Page

In [3]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html = urlopen(url) 
soup = BeautifulSoup(html, 'html.parser')

In [4]:
soup = soup.find_all("table")

In [5]:
# Scraping the headers

columns=[]
tables = soup

for table in tables:
    
    header_rows = table.find_all('tr')
    
    for hr in header_rows:
        headers = hr.find_all('th')
        #print(headers)
        
        if len(headers) > 1:
            for i in range(3):
                data = headers[i]
                columns.append(data.text.strip())

columns
            

['Postal Code', 'Borough', 'Neighborhood']

In [6]:
#Scraping the table data
tables = soup

postal_code=[]
borough =[]
neighborhood=[]

for table in tables:
    
    rows = table.find_all('tr')
    
    for r in rows:
        data = r.find_all('td')
        
        if len(data) > 1:
            
            d1 = data[0]
            postal_code.append(d1.text.strip())
            
            d2 = data[1]
            borough.append(d2.text.strip())
             
            if len(data) > 2:
                d3 = data[2]
                neighborhood.append(d3.text.strip())
    break                  

In [7]:
print(postal_code)

['M1A', 'M2A', 'M3A', 'M4A', 'M5A', 'M6A', 'M7A', 'M8A', 'M9A', 'M1B', 'M2B', 'M3B', 'M4B', 'M5B', 'M6B', 'M7B', 'M8B', 'M9B', 'M1C', 'M2C', 'M3C', 'M4C', 'M5C', 'M6C', 'M7C', 'M8C', 'M9C', 'M1E', 'M2E', 'M3E', 'M4E', 'M5E', 'M6E', 'M7E', 'M8E', 'M9E', 'M1G', 'M2G', 'M3G', 'M4G', 'M5G', 'M6G', 'M7G', 'M8G', 'M9G', 'M1H', 'M2H', 'M3H', 'M4H', 'M5H', 'M6H', 'M7H', 'M8H', 'M9H', 'M1J', 'M2J', 'M3J', 'M4J', 'M5J', 'M6J', 'M7J', 'M8J', 'M9J', 'M1K', 'M2K', 'M3K', 'M4K', 'M5K', 'M6K', 'M7K', 'M8K', 'M9K', 'M1L', 'M2L', 'M3L', 'M4L', 'M5L', 'M6L', 'M7L', 'M8L', 'M9L', 'M1M', 'M2M', 'M3M', 'M4M', 'M5M', 'M6M', 'M7M', 'M8M', 'M9M', 'M1N', 'M2N', 'M3N', 'M4N', 'M5N', 'M6N', 'M7N', 'M8N', 'M9N', 'M1P', 'M2P', 'M3P', 'M4P', 'M5P', 'M6P', 'M7P', 'M8P', 'M9P', 'M1R', 'M2R', 'M3R', 'M4R', 'M5R', 'M6R', 'M7R', 'M8R', 'M9R', 'M1S', 'M2S', 'M3S', 'M4S', 'M5S', 'M6S', 'M7S', 'M8S', 'M9S', 'M1T', 'M2T', 'M3T', 'M4T', 'M5T', 'M6T', 'M7T', 'M8T', 'M9T', 'M1V', 'M2V', 'M3V', 'M4V', 'M5V', 'M6V', 'M7V', 'M8V'

In [8]:
print(borough)

['Not assigned', 'Not assigned', 'North York', 'North York', 'Downtown Toronto', 'North York', 'Downtown Toronto', 'Not assigned', 'Etobicoke', 'Scarborough', 'Not assigned', 'North York', 'East York', 'Downtown Toronto', 'North York', 'Not assigned', 'Not assigned', 'Etobicoke', 'Scarborough', 'Not assigned', 'North York', 'East York', 'Downtown Toronto', 'York', 'Not assigned', 'Not assigned', 'Etobicoke', 'Scarborough', 'Not assigned', 'Not assigned', 'East Toronto', 'Downtown Toronto', 'York', 'Not assigned', 'Not assigned', 'Not assigned', 'Scarborough', 'Not assigned', 'Not assigned', 'East York', 'Downtown Toronto', 'Downtown Toronto', 'Not assigned', 'Not assigned', 'Not assigned', 'Scarborough', 'North York', 'North York', 'East York', 'Downtown Toronto', 'West Toronto', 'Not assigned', 'Not assigned', 'Not assigned', 'Scarborough', 'North York', 'North York', 'East York', 'Downtown Toronto', 'West Toronto', 'Not assigned', 'Not assigned', 'Not assigned', 'Scarborough', 'North

In [9]:
print(neighborhood)

['', '', 'Parkwoods', 'Victoria Village', 'Regent Park, Harbourfront', 'Lawrence Manor, Lawrence Heights', "Queen's Park, Ontario Provincial Government", '', 'Islington Avenue, Humber Valley Village', 'Malvern, Rouge', '', 'Don Mills', 'Parkview Hill, Woodbine Gardens', 'Garden District, Ryerson', 'Glencairn', '', '', 'West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale', 'Rouge Hill, Port Union, Highland Creek', '', 'Don Mills', 'Woodbine Heights', 'St. James Town', 'Humewood-Cedarvale', '', '', 'Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood', 'Guildwood, Morningside, West Hill', '', '', 'The Beaches', 'Berczy Park', 'Caledonia-Fairbanks', '', '', '', 'Woburn', '', '', 'Leaside', 'Central Bay Street', 'Christie', '', '', '', 'Cedarbrae', 'Hillcrest Village', 'Bathurst Manor, Wilson Heights, Downsview North', 'Thorncliffe Park', 'Richmond, Adelaide, King', 'Dufferin, Dovercourt Village', '', '', '', 'Scarborough Village', 'Fairview, Henry Farm, Oriole

#### Creating the Dataframe

In [10]:
df = pd.DataFrame({columns[0]:postal_code, columns[1]:borough, columns[2]:neighborhood})
print(df.shape)
df.head()

(180, 3)


Postal Code           Borough               Neighborhood
0         M1A      Not assigned                           
1         M2A      Not assigned                           
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Dataframe has 180 rows and 3 columns.

#### Dropping the rows with no assigned Borough

In [11]:
df = df[df['Borough'] != 'Not assigned']
print(df.shape)
df.head()

(103, 3)


Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

After deleting the rows with 'Not assigned' value in the Borough column, we have 103 rows left in the dataframe.

In [12]:
df.reset_index(inplace=True) #resetting the index since we deleted some rows.
df.drop(['index'], axis = 1, inplace=True)
df.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

#### Grouping the data by Postcode and Borough columns to merge any repeated Postal codes

In [13]:
df_final = df.groupby(['Postal Code', 'Borough'])['Neighborhood'].apply(', '.join).reset_index()
df_final

Postal Code           Borough  \
0           M1B       Scarborough   
1           M1C       Scarborough   
2           M1E       Scarborough   
3           M1G       Scarborough   
4           M1H       Scarborough   
5           M1J       Scarborough   
6           M1K       Scarborough   
7           M1L       Scarborough   
8           M1M       Scarborough   
9           M1N       Scarborough   
10          M1P       Scarborough   
11          M1R       Scarborough   
12          M1S       Scarborough   
13          M1T       Scarborough   
14          M1V       Scarborough   
15          M1W       Scarborough   
16          M1X       Scarborough   
17          M2H        North York   
18          M2J        North York   
19          M2K        North York   
20          M2L        North York   
21          M2M        North York   
22          M2N        North York   
23          M2P        North York   
24          M2R        North York   
25          M3A        North York   
26          M3B        North York   
27          M3C        North York   
28          M3H        North York   
29          M3J        North York   
..          ...               ...   
73          M6C              York   
74          M6E              York   
75          M6G  Downtown Toronto   
76          M6H      West Toronto   
77          M6J      West Toronto   
78          M6K      West Toronto   
79          M6L        North York   
80          M6M              York   
81          M6N              York   
82          M6P      West Toronto   
83          M6R      West Toronto   
84          M6S      West Toronto   
85          M7A  Downtown Toronto   
86          M7R       Mississauga   
87          M7Y      East Toronto   
88          M8V         Etobicoke   
89          M8W         Etobicoke   
90          M8X         Etobicoke   
91          M8Y         Etobicoke   
92          M8Z         Etobicoke   
93          M9A         Etobicoke   
94          M9B         Etobicoke   
95          M9C         Etobicoke   
96          M9L        North York   
97          M9M        North York   
98          M9N              York   
99          M9P         Etobicoke   
100         M9R         Etobicoke   
101         M9V         Etobicoke   
102         M9W         Etobicoke   

                                          Neighborhood  
0                                       Malvern, Rouge  
1               Rouge Hill, Port Union, Highland Creek  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6          Kennedy Park, Ionview, East Birchmount Park  
7                      Golden Mile, Clairlea, Oakridge  
8      Cliffside, Cliffcrest, Scarborough Village West  
9                          Birch Cliff, Cliffside West  
10   Dorset Park, Wexford Heights, Scarborough Town...  
11                                   Wexford, Maryvale  
12                                           Agincourt  
13             Clarks Corners, Tam O'Shanter, Sullivan  
14   Milliken, Agincourt North, Steeles East, L'Amo...  
15                       Steeles West, L'Amoreaux West  
16                                         Upper Rouge  
17                                   Hillcrest Village  
18                        Fairview, Henry Farm, Oriole  
19                                     Bayview Village  
20                            York Mills, Silver Hills  
21                             Willowdale, Newtonbrook  
22                         Willowdale, Willowdale East  
23                                     York Mills West  
24                         Willowdale, Willowdale West  
25                                           Parkwoods  
26                                           Don Mills  
27                                           Don Mills  
28     Bathurst Manor, Wilson Heights, Downsview North  

#### Checking for empty ('') and not assigned ('Not assigned') values in the Neighbourhood column

In [14]:
print((df_final['Neighborhood'] != '').value_counts())
print((df_final['Neighborhood'] != 'Not assigned').value_counts())

True    103
Name: Neighborhood, dtype: int64
True    103
Name: Neighborhood, dtype: int64


Since every value returned true, there are neither empty nor Not assigned neighborhoods in the Dataframe.

#### Printing the shape of the Dataframe

In [15]:
df_final.shape

(103, 3)

## Q.2

#### Getting the coordinates CSV file

In [16]:
df_ll = pd.read_csv('http://cocl.us/Geospatial_data')
df_ll.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

#### Merging it with the Neighborhood dataframe

In [17]:
df_1 = pd.merge(df_final,df_ll,on='Postal Code')
df_1.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

## Q.3

In [23]:
#!conda install -c conda-forge folium=0.5.0 --yes
import folium
import requests # library to handle requests
from pandas.io.json import json_normalize 
from geopy.geocoders import Nominatim

In [24]:
# Getting Torronto Coordinates

address = 'Toronto, Canada'
geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of the City of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of the City of Toronto are 43.6534817, -79.3839347.


In [73]:
# Creating a map
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_1['Latitude'], df_1['Longitude'], df_1['Borough'], df_1['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=4,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Exploring Neighborhoods of toronto

In [72]:
CLIENT_ID = '******' # your Foursquare ID
CLIENT_SECRET = '*******' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
#print('CLIENT_ID: ' + CLIENT_ID)
#print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:


#### Seperating the boroughs that contain the word Toronto 

In [27]:
df_3 = df_1[df_1['Borough'].str.contains('Toronto')]
df_3.reset_index(inplace=True) # resetting the index column
df_3.drop('index', axis=1, inplace=True)
df_3

/opt/conda/envs/Python36/lib/python3.6/site-packages/pandas/core/frame.py:3940: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


Postal Code           Borough  \
0          M4E      East Toronto   
1          M4K      East Toronto   
2          M4L      East Toronto   
3          M4M      East Toronto   
4          M4N   Central Toronto   
5          M4P   Central Toronto   
6          M4R   Central Toronto   
7          M4S   Central Toronto   
8          M4T   Central Toronto   
9          M4V   Central Toronto   
10         M4W  Downtown Toronto   
11         M4X  Downtown Toronto   
12         M4Y  Downtown Toronto   
13         M5A  Downtown Toronto   
14         M5B  Downtown Toronto   
15         M5C  Downtown Toronto   
16         M5E  Downtown Toronto   
17         M5G  Downtown Toronto   
18         M5H  Downtown Toronto   
19         M5J  Downtown Toronto   
20         M5K  Downtown Toronto   
21         M5L  Downtown Toronto   
22         M5N   Central Toronto   
23         M5P   Central Toronto   
24         M5R   Central Toronto   
25         M5S  Downtown Toronto   
26         M5T  Downtown Toronto   
27         M5V  Downtown Toronto   
28         M5W  Downtown Toronto   
29         M5X  Downtown Toronto   
30         M6G  Downtown Toronto   
31         M6H      West Toronto   
32         M6J      West Toronto   
33         M6K      West Toronto   
34         M6P      West Toronto   
35         M6R      West Toronto   
36         M6S      West Toronto   
37         M7A  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                        The Danforth West, Riverdale  43.679557 -79.352188  
2                      India Bazaar, The Beaches West  43.668999 -79.315572  
3                                     Studio District  43.659526 -79.340923  
4                                       Lawrence Park  43.728020 -79.388790  
5                                    Davisville North  43.712751 -79.390197  
6                  North Toronto West,  Lawrence Park  43.715383 -79.405678  
7                                          Davisville  43.704324 -79.388790  
8                         Moore Park, Summerhill East  43.689574 -79.383160  
9   Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
10                                           Rosedale  43.679563 -79.377529  
11                        St. James Town, Cabbagetown  43.667967 -79.367675  
12                               Church and Wellesley  43.665860 -79.383160  
13                          Regent Park, Harbourfront  43.654260 -79.360636  
14                           Garden District, Ryerson  43.657162 -79.378937  
15                                     St. James Town  43.651494 -79.375418  
16                                        Berczy Park  43.644771 -79.373306  
17                                 Central Bay Street  43.657952 -79.387383  
18                           Richmond, Adelaide, King  43.650571 -79.384568  
19  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
20           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
21                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
22                                           Roselawn  43.711695 -79.416936  
23    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                     University of Toronto, Harbord  43.662696 -79.400049  
26          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
27  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
28                                     Stn A PO Boxes  43.646435 -79.374846  
29             First Canadian Place, Underground city  43.648429 -79.382280  
30                                           Christie  43.669542 -79.422564  
31                       Dufferin, Dovercourt Village  43.669005 -

In [28]:
# Creating a map
map_tneighborhoods = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_1['Latitude'], df_3['Longitude'], df_3['Borough'], df_3['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=4,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_tneighborhoods)  
    
map_tneighborhoods

#### Exploring the first neighborhood

Getting the first neighborhood name

In [29]:
df_3.loc[0, 'Neighborhood']

'The Beaches'

Getting the Neighborhood latitude and longitude values.

In [30]:
neighborhood_latitude = df_3.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df_3.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df_3.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


Let's use Foursquare-api to fetch top-100 venues within 500 meters of 'The Beaches' Neighborhood.

In [31]:
radius = 500
LIMIT = 100

url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, neighborhood_latitude, neighborhood_longitude, VERSION, radius, LIMIT)

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ed26b14963d29001b4c2333'},
 'response': {'headerLocation': 'The Beaches',
  'headerFullLocation': 'The Beaches, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 6,
  'suggestedBounds': {'ne': {'lat': 43.680857404499996,
    'lng': -79.28682091449052},
   'sw': {'lat': 43.67185739549999, 'lng': -79.29924148550948}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bd461bc77b29c74a07d9282',
       'name': 'Glen Manor Ravine',
       'location': {'address': 'Glen Manor',
        'crossStreet': 'Queen St.',
        'lat': 43.67682094413784,
        'lng': -79.29394208780985,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.67682094413784,
          'lng': -79.29394208780985}],
        'distanc

In [32]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [33]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name         categories        lat        lng
0                   Glen Manor Ravine              Trail  43.676821 -79.293942
1  The Big Carrot Natural Food Market  Health Food Store  43.678879 -79.297734
2                 Grover Pub and Grub                Pub  43.679181 -79.297215
3                   Glen Stewart Park               Park  43.675278 -79.294647
4                       Upper Beaches       Neighborhood  43.680563 -79.292869

In [34]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

6 venues were returned by Foursquare.


#### Exploring all neighborhoods in Toronto

In [35]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [36]:
toronto_venues = getNearbyVenues(names=df_3['Neighborhood'],
                                   latitudes=df_3['Latitude'],
                                   longitudes=df_3['Longitude']
                                  )


The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West,  Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High

In [37]:
print(toronto_venues.shape)
toronto_venues.head()

(1616, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                   Glen Stewart Park       43.675278       -79.294647   
4                       Upper Beaches       43.680563       -79.292869   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3               Park  
4       Neighborhood

In [38]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     17   
CN Tower, King and Spadina, Railway Lands, Harb...                     14   
Central Bay Street                                                     62   
Christie                                                               17   
Church and Wellesley                                                   77   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        7   
Dufferin, Dovercourt Village                                           16   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
India Bazaar, The Beaches West                                         22   
Kensington Market, Chinatown, Grange Park                              58   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               44   
Moore Park, Summerhill East                                             3   
North Toronto West,  Lawrence Park                                     22   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            36   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                               91   
Rosedale                                                                4   
Roselawn                                                                3   
Runnymede, Swansea                                                     42   
St. James Town                                                         80   
St. James Town, Cabbagetown                                            42   
Stn A PO Boxes                                                         93   
Studio District                                                        41   
Summerhill West, Rathnelly, South Hill, Forest ...                     17   
The Annex, North Midtown, Yorkville                                    22   
The Beaches                                                             6   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             58   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      17   
CN Tower, King and Spadina, Railway Lands, Harb...                      14   
Central Bay Street                                                      62   
Christie                                                                17   
Church and Wellesley                                                    77   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              33   
Davisville North                    

#### Analyzing each Neighborhood

In [39]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

toronto_onehot.drop(columns=['Neighborhood'],inplace=True)

toronto_onehot.insert(loc=0, column='Neighborhood', value=toronto_venues['Neighborhood'])
toronto_onehot.head()

Neighborhood  Afghan Restaurant  Airport  Airport Food Court  \
0  The Beaches                  0        0                   0   
1  The Beaches                  0        0                   0   
2  The Beaches                  0        0                   0   
3  The Beaches                  0        0                   0   
4  The Beaches                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Toy / Game Store  Trail  Train Station  \
0             0         0  ...                 0      1              0   
1             0         0  ...                 0      0              0   
2             0         0  ...                 0      0              0   
3             0         0  ...                 0      0              0   
4             0         0  ...                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0         0            0              0            0  
1         0            0              0            0  
2         0            0              0            0  
3         0            0              0            0  
4         0            0              0            0  

[5 rows x 236 columns]

In [40]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Afghan Restaurant  \
0                                         Berczy Park           0.000000   
1        Brockton, Parkdale Village, Exhibition Place           0.000000   
2   Business reply mail Processing Centre, South C...           0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
4                                  Central Bay Street           0.000000   
5                                            Christie           0.000000   
6                                Church and Wellesley           0.012987   
7                      Commerce Court, Victoria Hotel           0.000000   
8                                          Davisville           0.000000   
9                                    Davisville North           0.000000   
10                       Dufferin, Dovercourt Village           0.000000   
11             First Canadian Place, Underground city           0.000000   
12    Forest Hill North & West, Forest Hill Road Park           0.000000   
13                           Garden District, Ryerson           0.000000   
14  Harbourfront East, Union Station, Toronto Islands           0.000000   
15                      High Park, The Junction South           0.000000   
16                     India Bazaar, The Beaches West           0.000000   
17          Kensington Market, Chinatown, Grange Park           0.000000   
18                                      Lawrence Park           0.000000   
19                           Little Portugal, Trinity           0.000000   
20                        Moore Park, Summerhill East           0.000000   
21                 North Toronto West,  Lawrence Park           0.000000   
22                             Parkdale, Roncesvalles           0.000000   
23        Queen's Park, Ontario Provincial Government           0.000000   
24                          Regent Park, Harbourfront           0.000000   
25                           Richmond, Adelaide, King           0.000000   
26                                           Rosedale           0.000000   
27                                           Roselawn           0.000000   
28                                 Runnymede, Swansea           0.000000   
29                                     St. James Town           0.000000   
30                        St. James Town, Cabbagetown           0.000000   
31                                     Stn A PO Boxes           0.000000   
32                                    Studio District           0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
34                The Annex, North Midtown, Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                       The Danforth West, Riverdale           0.000000   
37           Toronto Dominion Centre, Design Exchange           0.000000   
38                     University of Toronto, Harbord           0.000000   

     Airport  Airport Food Court  Airport Lounge  Airport Service  \
0   0.000000            0.000000        0.000000         0.000000   
1   0.000000            0.000000        0.000000         0.000000   
2   0.000000            0.000000        0.000000         0.000000   
3   0.071429            0.071429        0.142857         0.142857   
4   0.000000            0.000000        0.000000         0.000000   
5   0.000000            0.000000        0.000000         0.000000   
6   0.000000            0.000000        0.000000         0.000000   
7   0.000000            0.000000        0.000000         0.000000   
8   0.000000            0.000000        0.000000         0.000000   
9   0.000000            0.000000        0.000000         0.000000   
10  0.000000            0.000000        0.000000         0.000000   
11  0.000000            0.000000        0.000000         0.000000   
12  0.000000            0.000000        0.000000         0.000000   
13  0.000000            0.000000  

#### Printing each neighborhood along with the top 5 most common venues

In [41]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2            Beer Bar  0.03
3  Seafood Restaurant  0.03
4                 Pub  0.03


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1  Breakfast Spot  0.09
2       Nightclub  0.09
3     Coffee Shop  0.09
4          Bakery  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                venue  freq
0  Light Rail Station  0.12
1         Yoga Studio  0.06
2       Garden Center  0.06
3          Comic Shop  0.06
4                Park  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.14
1   Airport Service  0.14
2  Airport Terminal  0.14
3          Boutique  0.07
4           Airport  0.07


----Central Bay Street----
                 venue  freq
0          Coffee S

In [62]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar                  Café                   Pub   
1        Breakfast Spot             Nightclub           Coffee Shop   
2           Yoga Studio         Auto Workshop            Skate Park   
3       Airport Service      Airport Terminal         Boat or Ferry   
4    Italian Restaurant        Sandwich Place                  Café   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Bakery            Restaurant           Cheese Shop   
1  Furniture / Home Store         Burrito Place    Italian Restaurant   
2                 Brewery                   Spa        Farmers Market   
3         Harbor / Marina      Sculpture Garden                 Plane   
4         Bubble Tea Shop          Burger Joint   Japanese Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Beer Bar    Seafood Restaurant                 Lounge  
1               Stadium          Intersection             Restaurant  
2  Fast Food Restaurant         Burrito Place                Butcher  
3              Boutique                   Bar     Airport Food Court  
4           Salad Place            Restaurant       Ramen Restaurant

#### K-means Clustering

In [64]:
from sklearn.cluster import KMeans


# set number of clusters
kclusters = 4

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0, init="k-means++", n_init=12).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

#### Adding Clustered Labels to the dataframe

In [65]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_3

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() 

Postal Code          Borough                    Neighborhood   Latitude  \
0         M4E     East Toronto                     The Beaches  43.676357   
1         M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2         M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3         M4M     East Toronto                 Studio District  43.659526   
4         M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0                   Pub                  Park   
1 -79.352188               0      Greek Restaurant    Italian Restaurant   
2 -79.315572               0  Fast Food Restaurant        Sandwich Place   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.388790               1                  Park              Bus Line   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Coffee Shop     Health Food Store                 Trail   
1           Coffee Shop            Restaurant        Ice Cream Shop   
2           Pizza Place      Sushi Restaurant            Steakhouse   
3   American Restaurant                Bakery               Brewery   
4           Swim School           Yoga Studio      Department Store   

    6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0             Yoga Studio      Department Store                 Dessert Shop   
1  Furniture / Home Store           Yoga Studio                 Liquor Store   
2          Ice Cream Shop     Fish & Chips Shop                      Brewery   
3               Gastropub           Yoga Studio                  Fish Market   
4    Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0                 Diner         Discount Store  
1                   Spa              Juice Bar  
2     Food & Drink Shop                    Pub  
3           Pizza Place              Pet Store  
4            Donut Shop       Doner Restaurant

#### Plotting Clustered neighborhoods data

In [66]:
import matplotlib.cm as cm
import matplotlib.colors as colors


# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Examining the clusters:

<b>Cluster-0 </b> :

In [67]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + [2]+ list(range(5, toronto_merged.shape[1]))]]

Borough                                       Neighborhood  \
0       East Toronto                                        The Beaches   
1       East Toronto                       The Danforth West, Riverdale   
2       East Toronto                     India Bazaar, The Beaches West   
3       East Toronto                                    Studio District   
5    Central Toronto                                   Davisville North   
6    Central Toronto                 North Toronto West,  Lawrence Park   
7    Central Toronto                                         Davisville   
8    Central Toronto                        Moore Park, Summerhill East   
9    Central Toronto  Summerhill West, Rathnelly, South Hill, Forest...   
11  Downtown Toronto                        St. James Town, Cabbagetown   
12  Downtown Toronto                               Church and Wellesley   
13  Downtown Toronto                          Regent Park, Harbourfront   
14  Downtown Toronto                           Garden District, Ryerson   
15  Downtown Toronto                                     St. James Town   
16  Downtown Toronto                                        Berczy Park   
17  Downtown Toronto                                 Central Bay Street   
18  Downtown Toronto                           Richmond, Adelaide, King   
19  Downtown Toronto  Harbourfront East, Union Station, Toronto Islands   
20  Downtown Toronto           Toronto Dominion Centre, Design Exchange   
21  Downtown Toronto                     Commerce Court, Victoria Hotel   
24   Central Toronto                The Annex, North Midtown, Yorkville   
25  Downtown Toronto                     University of Toronto, Harbord   
26  Downtown Toronto          Kensington Market, Chinatown, Grange Park   
27  Downtown Toronto  CN Tower, King and Spadina, Railway Lands, Har...   
28  Downtown Toronto                                     Stn A PO Boxes   
29  Downtown Toronto             First Canadian Place, Underground city   
30  Downtown Toronto                                           Christie   
31      West Toronto                       Dufferin, Dovercourt Village   
32      West Toronto                           Little Portugal, Trinity   
33      West Toronto       Brockton, Parkdale Village, Exhibition Place   
34      West Toronto                      High Park, The Junction South   
35      West Toronto                             Parkdale, Roncesvalles   
36      West Toronto                                 Runnymede, Swansea   
37  Downtown Toronto        Queen's Park, Ontario Provincial Government   
38      East Toronto  Business reply mail Processing Centre, South C...   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0                0                   Pub                  Park   
1                0      Greek Restaurant    Italian Restaurant   
2                0  Fast Food Restaurant        Sandwich Place   
3                0                  Café           Coffee Shop   
5                0                   Gym        Breakfast Spot   
6                0        Clothing Store           Coffee Shop   
7                0        Sandwich Place          Dessert Shop   
8                0            Playground           Summer Camp   
9                0           Coffee Shop                  Bank   
11               0           Coffee Shop                  Café   
12               0           Coffee Shop   Japanese Restaurant   
13               0           Coffee Shop                   Pub   
14               0        Clothing Store           Coffee Shop   
15               0           Coffee Shop                  Café   
16               0           Coffee Shop          Cocktail Bar   
17               0           Coffee Shop    Italian Restaurant   
18               0           Coffee Shop                  Café   
19               0           Coffee Shop              Aquarium   
20               0           Coffee Shop                 H

<b> Cluster-1 </b>

In [68]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + [2] + list(range(5, toronto_merged.shape[1]))]]

Borough   Neighborhood  Cluster Labels 1st Most Common Venue  \
4  Central Toronto  Lawrence Park               1                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4              Bus Line           Swim School           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4      Department Store  Ethiopian Restaurant     Electronics Store   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4  Eastern European Restaurant            Donut Shop       Doner Restaurant

<b> Cluster-2 </b>

In [69]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + [2] + list(range(5, toronto_merged.shape[1]))]]

Borough                                     Neighborhood  \
10  Downtown Toronto                                         Rosedale   
23   Central Toronto  Forest Hill North & West, Forest Hill Road Park   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
10               2                  Park            Playground   
23               2                  Park         Jewelry Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
10                 Trail           Yoga Studio         Deli / Bodega   
23                 Trail      Sushi Restaurant           Yoga Studio   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
10  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   
23      Department Store  Ethiopian Restaurant            Electronics Store   

          9th Most Common Venue 10th Most Common Venue  
10                   Donut Shop       Doner Restaurant  
23  Eastern European Restaurant             Donut Shop

<b> Cluster-3 </b>

In [70]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + [2] + list(range(5, toronto_merged.shape[1]))]]

Borough Neighborhood  Cluster Labels 1st Most Common Venue  \
22  Central Toronto     Roselawn               3        Ice Cream Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22          Home Service                Garden         Deli / Bodega   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22           Event Space  Ethiopian Restaurant     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22  Eastern European Restaurant            Donut Shop       Doner Restaurant